src/crimson/os/seastore/omap_manager/btree/omap_btree_node.h

```diff
@@ -0,0 +1,142 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+#pragma once
+
+#include <string>
+#include <vector>
+
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "crimson/common/log.h"
+#include "crimson/os/seastore/seastore_types.h"
+#include "crimson/os/seastore/transaction_manager.h"
+#include "crimson/os/seastore/omap_manager.h"
+#include "crimson/os/seastore/omap_manager/btree/omap_types.h"
+
+namespace crimson::os::seastore::omap_manager{
+
+struct omap_context_t {
+  omap_root_t &omap_root;
+  TransactionManager &tm;
+  Transaction &t;
+};
+
+struct OMapNode : LogicalCachedExtent {
+  using OMapNodeRef = TCachedExtentRef<OMapNode>;
+  OMapNodeRef parent = nullptr;
+
+  OMapNode(ceph::bufferptr &&ptr) : LogicalCachedExtent(std::move(ptr)) {}
+  OMapNode(const OMapNode &other)
+  : LogicalCachedExtent(other) {}
+
+  using get_value_ertr = OMapManager::omap_get_value_ertr;
+  using get_value_ret = OMapManager::omap_get_value_ret;
+  virtual get_value_ret get_value(omap_context_t oc, const std::string &key) = 0;
+
+  using insert_ertr = OMapManager::omap_set_key_ertr;
+  using insert_ret = OMapManager::omap_set_key_ret;
+  virtual insert_ret insert(omap_context_t oc, std::string &key, std::string &value) = 0;
+
+  using rm_key_ertr = OMapManager::omap_rm_key_ertr;
+  using rm_key_ret = OMapManager::omap_rm_key_ret;
+  virtual rm_key_ret rm_key(omap_context_t oc, const std::string &key) = 0;
+
+  using list_keys_ertr = OMapManager::omap_list_keys_ertr;
+  using list_keys_ret = list_keys_ertr::future<>;
+  virtual list_keys_ret list_keys(omap_context_t oc, std::vector<std::string> &result) = 0;
```

Comment:
We need to support partial listing.  The interface should probably be something like:
```
struct list_keys_result_t {
  std::vector<std::string> keys;
  std::string next;
};
using list_keys_ret = list_keys_ertr::future<list_keys_result_t>;
virtual list_keys_ret list_keys(omap_context_t oc, std::string next) = 0;
```

Reply:
@athanatos  in virtual list_keys_ret list_keys(omap_context_t oc, std::string next, size_t max_result_size) = 0;  the next means start key, right?   do you mean we give a start key(next) and  max_result_size , so list keys from the start key and until the keys number =   max_result_size ? and in list_keys_result_t struct, the next mean the one  after start key + max_result_size ?

Reply:
Yeah, that sounds right to me.